tree/ntuple/v7/src/RFieldMeta.cxx

```diff
@@ -186,39 +217,134 @@ std::size_t ROOT::Experimental::RClassField::AppendImpl(const void *from)
 
 void ROOT::Experimental::RClassField::ReadGlobalImpl(NTupleSize_t globalIndex, void *to)
 {
+   for (const auto &[_, si] : fStagingItems) {
+      CallReadOn(*si.fField, globalIndex, fStagingArea.get() + si.fOffset);
+   }
    for (unsigned i = 0; i < fSubFields.size(); i++) {
       CallReadOn(*fSubFields[i], globalIndex, static_cast<unsigned char *>(to) + fSubFieldsInfo[i].fOffset);
    }
 }
 
 void ROOT::Experimental::RClassField::ReadInClusterImpl(RNTupleLocalIndex localIndex, void *to)
 {
+   for (const auto &[_, si] : fStagingItems) {
+      CallReadOn(*si.fField, localIndex, fStagingArea.get() + si.fOffset);
+   }
    for (unsigned i = 0; i < fSubFields.size(); i++) {
       CallReadOn(*fSubFields[i], localIndex, static_cast<unsigned char *>(to) + fSubFieldsInfo[i].fOffset);
    }
 }
 
-void ROOT::Experimental::RClassField::BeforeConnectPageSource(Internal::RPageSource &pageSource)
+ROOT::Experimental::DescriptorId_t
+ROOT::Experimental::RClassField::LookupMember(const RNTupleDescriptor &desc, std::string_view memberName,
+                                              DescriptorId_t classFieldId)
 {
-   // This can happen for added base classes or non-simple members.
-   if (GetOnDiskId() == kInvalidDescriptorId) {
-      return;
+   auto idSourceMember = desc.FindFieldId(memberName, classFieldId);
+   if (idSourceMember != kInvalidDescriptorId)
+      return idSourceMember;
+
+   for (const auto &subFieldDesc : desc.GetFieldIterable(classFieldId)) {
+      const auto subFieldName = subFieldDesc.GetFieldName();
+      if (subFieldName.length() > 2 && subFieldName[0] == ':' && subFieldName[1] == '_') {
+         idSourceMember = LookupMember(desc, memberName, subFieldDesc.GetId());
+         if (idSourceMember != kInvalidDescriptorId)
+            return idSourceMember;
+      }
    }
-   // Gather all known sub fields in the descriptor.
+
+   return kInvalidDescriptorId;
+}
+
+void ROOT::Experimental::RClassField::PrepareStagingArea(const std::vector<const TSchemaRule *> &rules,
+                                                         const RNTupleDescriptor &desc,
+                                                         const RFieldDescriptor &classFieldDesc)
+{
+   auto clName = classFieldDesc.GetTypeName();
+   if (classFieldDesc.GetTypeVersion() != GetTypeVersion()) {
+      TClass::GetClass(classFieldDesc.GetTypeName().c_str())->GetStreamerInfo(classFieldDesc.GetTypeVersion());
+      clName += std::string("@@") + std::to_string(classFieldDesc.GetTypeVersion());
+   }
+   auto cl = TClass::GetClass(clName.c_str());
```

Comment:
Finding the `@@` class relies on the related `StreamerInfo` to have been build, is that a guarantee here?

Reply:
I thought that the line before, calling `TClass::GetClass(<class name>)->GetStreamerInfo(<version used in @@>)` is enough. Is another call necessary?

Reply:
Yes, it should be enough.